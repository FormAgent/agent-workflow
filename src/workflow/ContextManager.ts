//上下文管理器
// 上下文用于任务之间的数据共享，并动态更新
import type { TaskInput } from "./Task";

export class ContextManager {
  private context: TaskInput = {};

  // 获取上下文中的数据
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  get(key: string): any {
    return this.context[key];
  }

  // 设置上下文中的数据
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  set(key: string, value: any): void {
    this.context[key] = value;
  }

  // 获取完整上下文
  getAll(): TaskInput {
    return this.context;
  }

  // 清除上下文
  clear(): void {
    this.context = {};
  }
}
