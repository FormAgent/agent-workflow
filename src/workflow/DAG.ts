import type { TaskExecutor } from './TaskExecutor';
import type { Task, TaskRegistry } from './TaskRegistry';

export interface DAGTask extends Task {
  dependsOn?: string[]; // 前置任务
}

export interface DAG {
  tasks: DAGTask[];
}

// biome-ignore lint/complexity/noStaticOnlyClass: <explanation>
export class DAGParser {
  static getExecutionOrderWithLevels(dag: DAG): { level: number; tasks: string[] }[] {
    const graph = new Map<string, string[]>(); // 任务图，映射每个任务到其依赖它的任务
    const inDegree = new Map<string, number>(); // 入度表，映射每个任务到其依赖的任务数量
    const levels = new Map<string, number>(); // 每个任务的层级

    // 初始化入度表和图
    for (const task of dag.tasks) {
      inDegree.set(task.name, task.dependsOn ? task.dependsOn.length : 0);
      // 初始化图，映射每个依赖任务到依赖它的任务
      if (task.dependsOn) {
        for (const dependency of task.dependsOn) {
          if (!graph.has(dependency)) {
            graph.set(dependency, []);
          }
          graph.get(dependency)?.push(task.name);
        }
      }
    }

    // 初始化队列，将入度为0的任务加入队列
    const queue: string[] = [];
    for (const [taskName, degree] of inDegree.entries()) {
      if (degree === 0) {
        queue.push(taskName);
        levels.set(taskName, 0);
      }
    }

    const result: { level: number; tasks: string[] }[] = [];
    while (queue.length > 0) {
      const taskName = queue.shift();
      if (!taskName) {
        throw new Error("Unexpected empty queue");
      }
      const currentLevel = levels.get(taskName);
      if (currentLevel === undefined) {
        throw new Error(`Level for task "${taskName}" not found`);
      }

      if (!result[currentLevel]) {
        result[currentLevel] = { level: currentLevel, tasks: [] };
      }
      result[currentLevel].tasks.push(taskName);

      // 遍历所有依赖于当前任务的任务
      for (const dependent of graph.get(taskName) || []) {
        const dependentInDegree = inDegree.get(dependent);
        if (dependentInDegree !== undefined) {
          inDegree.set(dependent, dependentInDegree - 1);
        }
        if (inDegree.get(dependent) === 0) {
          queue.push(dependent);
          levels.set(dependent, currentLevel + 1);
        }
      }
    }

    // 检查循环依赖
    if (levels.size !== dag.tasks.length) {
      throw new Error("Cyclic dependency detected in the task graph");
    }

    // 不再反转结果，保持从低层级到高层级的顺序
    return result;
  }
}


export class DAGWorkflowEngine {
  private taskRegistry: TaskRegistry;
  private executor: TaskExecutor;

  constructor(taskRegistry: TaskRegistry, executor: TaskExecutor) {
    this.taskRegistry = taskRegistry;
    this.executor = executor;
  }

  // 根据任务依赖图执行工作流
  async run(dag: DAG): Promise<void> {
    const levels = DAGParser.getExecutionOrderWithLevels(dag); // 按层解析任务
    for (const { level, tasks } of levels) {
      console.log(`Executing tasks at level ${level}:`, tasks);

      // 同一层级任务并行执行
      await Promise.all(
        tasks.map(async (taskName) => {
          const task = this.taskRegistry.getTask(taskName);
          if (!task) throw new Error(`Task "${taskName}" not found`);
          await this.executor.execute(task);
        })
      );
    }
    console.log("Workflow completed.");
  }
}

