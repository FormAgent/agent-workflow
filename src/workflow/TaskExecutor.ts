//任务执行器
// 任务执行器用于调用任务逻辑，并将结果更新到上下文中
import type { ContextManager } from './ContextManager';
import type { Task } from './Task';
import { ZodSchema } from 'zod';

export class TaskExecutor {
  private contextManager: ContextManager;

  constructor(contextManager: ContextManager) {
    this.contextManager = contextManager;
  }

  async execute(task: Task): Promise<void> {
    let input = this.contextManager.getAll(); // 获取上下文

    // Validate input using inputSchema if defined
    if (task.inputSchema) {
      try {
        input = await task.inputSchema.parseAsync(input);
      } catch (error) {
        console.error('Input validation failed:', error);
        throw error;
      }
    }

    let output = await task.execute(input); // 执行任务

    // Validate output using outputSchema if defined
    if (task.outputSchema) {
      try {
        output = await task.outputSchema.parseAsync(output);
      } catch (error) {
        console.error('Output validation failed:', error);
        throw error;
      }
    }

    for (const [key, value] of Object.entries(output)) {
      this.contextManager.set(key, value); // 更新上下文
    }
    console.log(`Task ${task} executed. Output:`, output);
  }

  getContext(): ContextManager {
    return this.contextManager;
  }
}
