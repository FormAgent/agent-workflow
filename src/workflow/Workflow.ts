// 工作流引擎
// 工作流引擎负责解析任务链并调度任务
// 支持条件分支，我们需要让工作流的执行路径根据任务的输出或上下文状态动态调整。这种能力可以用于处理多样化的输入需求，如：
// 根据意图选择不同的任务链。
// 根据任务结果动态跳转到不同的后续任务。
// 核心逻辑
// 条件节点

// 条件分支在工作流中用一个特殊节点表示。
// 节点包含条件逻辑（如函数）以及对应的后续任务链。
// 动态执行路径

// 解析条件节点时，根据上下文或任务输出选择合适的分支执行。
// 上下文依赖

// 条件分支可以访问并使用上下文数据，确保动态性和灵活性。

import type { ContextManager } from "./ContextManager";
import type { TaskExecutor } from "./TaskExecutor";
import type { Task } from "./Task";

// 条件分支
export interface ConditionBranch {
  condition: (context: ContextManager) => boolean; // 条件函数
  next: Task | Task[]; // 满足条件时的任务链
}

export interface ConditionNode {
  branches: ConditionBranch[]; // 所有分支
  default?: Task | Task[]; // 默认任务链（当所有条件都不满足时）
}

// 支持顺序,并行任务以及条件分支
type WorkflowStep = Task | Task[] | ConditionNode;

export interface WorkflowDefinition {
  steps: WorkflowStep[];
}

export class WorkflowEngine {
  private executor: TaskExecutor;

  constructor(executor: TaskExecutor) {
    this.executor = executor;
  }

  async run(workflow: WorkflowDefinition): Promise<void> {
    for (const step of workflow.steps) {
      if (Array.isArray(step)) {
        // 并行任务
        await Promise.all(step.map(task => this.runTask(task)));
      } else if (this.isConditionNode(step)) {
        // 条件分支
        await this.handleConditionNode(step as ConditionNode);
      } else {
        // 顺序任务
        await this.runTask(step);
      }
    }
  }

  private async runTask(task: Task): Promise<void> {
    await this.executor.execute(task);
  }

  private async handleConditionNode(node: ConditionNode): Promise<void> {
    const context = this.executor.getContext(); // 获取当前上下文
    for (const branch of node.branches) {
      if (branch.condition(context)) {
        console.log("Condition met, executing branch:", branch.next);
        await this.executeNext(branch.next);
        return;
      }
    }

    // 如果所有条件都不满足，执行默认分支
    if (node.default) {
      console.log("No conditions met, executing default branch:", node.default);
      await this.executeNext(node.default);
    }
  }

  private async executeNext(next: Task | Task[]): Promise<void> {
    if (Array.isArray(next)) {
      await Promise.all(next.map(task => this.runTask(task)));
    } else {
      await this.runTask(next);
    }
  }

  private isConditionNode(step: WorkflowStep): step is ConditionNode {
    return typeof step === "object" && "branches" in step;
  }
}
