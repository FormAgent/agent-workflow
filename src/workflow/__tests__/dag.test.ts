import { ContextManager } from '../ContextManager';
import { DAG, DAGParser } from '../DAG';
import { TaskExecutor } from '../TaskExecutor';
import type { TaskInput, TaskOutput, TaskRegistry } from '../TaskRegistry';
import { WorkflowEngine } from '../Workflow';

const execute = async (input: TaskInput): Promise<TaskOutput> => {
  return {};
};

describe('AIAgent', () => {
  beforeEach(() => {});

  test('dag1', async () => {
    const dag1: DAG = {
      tasks: [
        { name: 'TaskA', execute },
        { name: 'TaskB', dependsOn: ['TaskA'], execute },
        { name: 'TaskC', dependsOn: ['TaskB'], execute },
        { name: 'TaskD', dependsOn: ['TaskA'], execute },
      ],
    };
    const resposne = DAGParser.getExecutionOrderWithLevels(dag1);
    console.log(resposne); 
    expect(resposne.length).toBe(3);
    expect(resposne[0].level).toBe(0);
    expect(resposne[0].tasks).toEqual(['TaskA']);
    expect(resposne[1].level).toBe(1);
    expect(resposne[1].tasks).toEqual(['TaskB', 'TaskD']);
    expect(resposne[2].level).toBe(2);
    expect(resposne[2].tasks).toEqual(['TaskC']);
  });

  test('dag2', async () => {
    const dag2: DAG = {
      tasks: [
        { name: "TaskA", execute },
        { name: "TaskB", dependsOn: ["TaskA"], execute },
        { name: "TaskC", dependsOn: ["TaskA"], execute },
        { name: "TaskD", dependsOn: ["TaskB", "TaskC"], execute },
      ],
    };
    const resposne = DAGParser.getExecutionOrderWithLevels(dag2);
    console.log(resposne); 
    expect(resposne.length).toBe(3);
    expect(resposne[0].level).toBe(0);
    expect(resposne[0].tasks).toEqual(['TaskA']);
    expect(resposne[1].level).toBe(1);
    expect(resposne[1].tasks).toEqual(['TaskB', 'TaskC']);
    expect(resposne[2].level).toBe(2);
    expect(resposne[2].tasks).toEqual(['TaskD']);
  });

  test('dag3', async () => {
    const dag3: DAG = {
      tasks: [
        { name: "TaskA", execute },
        { name: "TaskB" , execute},
        { name: "TaskC", execute },
      ],
    };
    const resposne = DAGParser.getExecutionOrderWithLevels(dag3);
    console.log(resposne); 
    expect(resposne.length).toBe(1);
    expect(resposne[0].level).toBe(0);
    expect(resposne[0].tasks).toEqual(['TaskA','TaskB','TaskC']);
  });

  test('dag4', async () => {
    const dag4: DAG = {
      tasks: [
        { name: "TaskA", dependsOn: ["TaskC"], execute },
        { name: "TaskB", dependsOn: ["TaskA"], execute },
        { name: "TaskC", dependsOn: ["TaskB"], execute },
      ],
    };
    expect(() => {
    const resposne = DAGParser.getExecutionOrderWithLevels(dag4);
    console.log(resposne); 
    }).toThrow("Cyclic dependency detected in the task graph");
  });

});
