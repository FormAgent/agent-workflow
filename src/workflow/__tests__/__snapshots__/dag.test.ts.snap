// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AIAgent dag4: Cyclic dependency detected in the task graph 1`] = `"Cyclic dependency detected in the task graph"`;

exports[`AIAgent dag4: a Cyclic dependency detected in the task graph 1`] = `"Cyclic dependency detected in the task graph"`;

exports[`AIAgent dag4: aCyclic dependency detected in the task graph 1`] = `"Cyclic dependency detected in the task graph"`;
